velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -75,53 +75,64 @@ void PageReader::seekToPage(int64_t row) {
   }
 }
 
+// The size of Parquet Page Header remains unknown until the Header is parsed
+// because it is in Thrift format. To read Parquet Page Header, we are using an
+// adaptive approach to guess the Page Header size.
+
+// TODO: Refactor the inputStreams to have a general interface where read() can
+// be applied on different chunks of memory.
 PageHeader PageReader::readPageHeader(int64_t remainingSize) {
   // Note that sizeof(PageHeader) may be longer than actually read
   std::shared_ptr<thrift::ThriftBufferedTransport> transport;
   std::unique_ptr<apache::thrift::protocol::TCompactProtocolT<
       thrift::ThriftBufferedTransport>>
       protocol;
-  char copy[sizeof(PageHeader)];
-  bool wasInBuffer = false;
-  if (bufferEnd_ == bufferStart_) {
-    const void* buffer;
-    int32_t size;
-    inputStream_->Next(&buffer, &size);
-    bufferStart_ = reinterpret_cast<const char*>(buffer);
-    bufferEnd_ = bufferStart_ + size;
-  }
-  if (bufferEnd_ - bufferStart_ >= sizeof(PageHeader)) {
```

Comment:
Any way to keep this logic to avoid copying to the buffer?  This can cause performance regression in hot path.

Reply:
Currently no, because the page header might be on multiple `InputStream`. And the thrift reading can only read adjacent bytes. This can be avoided after refactoring the read() by adding a virtual layer to handle the addresses.

Reply:
I mean we can first try reading the buffer returned by the stream.  This is the most common case and copy should be avoided here.  After that fails, we can retry with the copy.  The initial size could be `bufferEnd_ - bufferStart_` here, and only that fails, we retry with `bufferEnd_ - bufferStart_ + pageHeaderBufferSize` and so on.